start on micro quark

ztatlock opened the pull request:
Let's try to build "micro Quark" in this branch.

It should be a super stripped down version. It doesn't even need to use WebKit.

The main thing we want to figure out is how the trace spec looks and how we can prove things about it.

@Ptival can you take ownership of this? Let's meet this morning to get started on it.

Commit:
start on micro quark

ztatlock commented:
@Ptival What if we make the component type correspond to the component's executable? e.g.

``` c

Init {
  spawn("UserInput");
  spawn("AddressBar");
}

Exchange {
  UserInput {
     ... handlers ...
  }
  AddressBar {
     ... handlers ...
  }
}
```

Ptival commented:
Is it the case that components are always created through an executable though? (like, for instance, how are tabs created?)

ztatlock commented:
Well, we can do things however we like.

But yes, right now I think all components will be created by spawn.

For example, tab creation could be like:

``` c
Messages {
  AddTab(str);
}

Init {
  spawn("UserInput");
  spawn("AddressBar");
}

Exchange {
  UserInput {
    AddTab(url) {
      spawn("Tab");
      // TODO we need some way to associate url with this tab
    }
     ... handlers ...
  }
  AddressBar {
     ... handlers ...
  }
  Tab {
    ... handlers ...
  }
}
```

Perhaps we should add a "Components" section to the DSL where the user defines all the paths to component executables and gives them names? e.g.

``` c
Components {
  UserInput := "user-input.py";
  Tab = "webkit-tab.py";
  AddressBar = "addr-bar";
}
```

Definitely feels crufty, but it may ease code generation for now...

Ptival commented:
I have a small problem figuring out the extracted code.

The thing is, I would have wished to define the state as a record:

``` ocaml
Record kstate :=
{ (* here the usual *)
; userinput : list chan * handler
; tabs : list chan * handler
}.
```

But with handler defined as:

``` ocaml
Definition handler := forall (c : chan) (s : kstate),
  STsep (kstate_inv kst * [In c (components kst)]) (fun (kst' : kstate) => kstate_inv kst').
```

But this creates all sort of inter-dependencies, since it refers both to kstate and kstate_inv.

Any thought?

ztatlock commented:
What happens if we just move all the Trace predicates (kstate_inv included) to earlier in the extracted code? Will that work?

Ptival commented:
No, the problem is not quite the order of things as much as how they are defined.

We can't have the state holding handlers whose type mentions an invariant over such states.

Either we loosen up the definition of handlers, so that they are not about holding an invariant over states, or we keep things separate somehow...

ztatlock commented:
One solution: don't put handlers in the kstate.

Instead make a component type type:

``` ocaml
Record kstate :=
{ (* here the usual *)
; components : list (component_t * list chan)
; ...
}.
```

Then have a dispatcher call the right handler based on the component type.

That way, handler can have the type you want and we break the dependency.

Does that sound reasonable?

Ptival commented:
That sounds reasonable.

ztatlock commented:
@Ptival Should I look into that bug you mentioned from the stateful branch?

 What exactly wasn't working?

Ptival committed:
Merge pull request #5 from UCSD-PL/stateful

Merging the first step towards Micro-Quark

Ptival commented:
So, it turns out that this is _the_ pull request from u-quark into master, and no other pull request from the same branch can exist at one time.

Let's try and create a pull request from stateful to master and see how this reacts! :)

Ptival committed:
Merge pull request #7 from UCSD-PL/components

components are now ready and working

Ptival committed:
Merge branch 'components' into u-quark

ztatlock commented:
You've been making great progress on this!

Ptival commented:
``` bash
State {
  currenttab : fdesc ;
  renderer   : fdesc ;
}

Components {
  Renderer  "renderer.py"  ;
  Tab       "tab.py"       ;
  UserInput "userinput.py" ;
}

Messages {
  Dummy(str);
  Input(str);
  Tab(str);
  Go(str);
  HTML(str);
}

Init {
  spawn(Renderer);
  spawn(Tab);
  spawn(UserInput);
}

Exchange(c) {
  Renderer {
    # A dummy message is sent on startup so that the chan can be stored
    Dummy(s) {
      renderer := c;
    }
  }

  Tab {
    # A dummy message is sent on startup so that the chan can be stored
    Dummy(s) {
      currenttab := c;
    }

    HTML(s) {
      # TODO: Check that it's the current tab
      send(renderer, HTML(s));
    }
  }

  UserInput {
    Input(s) {
      send(currenttab, Go(s));
    }
  }
}
```

What think?

ztatlock commented:
```
State {
  screen : fdesc;
  cur_tab : fdesc;
}

Components {
  Tab       "tab.py";
  Screen    "screen.py";
  UserInput "user-input.py";
}

Messages {
  Hello();
  Input(str);
  Wget(str);
  HTML(str, fdesc);
  Display(fdesc); 
}

Init {
  spawn(Tab);
  spawn(Screen);
  spawn(UserInput);
}

Exchange(c) {
  Tab {
    Hello() {
      cur_tab := c;
    }
    Wget(url) {
      f = call("wget.py");
      send(c, HTML(url, f));
    }
    Display(bitmap) {
      # TODO check that this is from cur tab
      send(screen, Display(bitmap)); 
    } 
  }

  Screen {
    Hello() {
      screen := c;
    }
  } 

  UserInput {
    Input(i) {
      send(cur_tab, Input(i));
    }
  }
} 
```

How about this?

Do the differences make sense?

ztatlock commented:
ooops...

Ptival commented:
Hmm, I need to add a Chan type.

@ztatlock Could you describe the changes needed in genC.ml and the subsequent changes needed on the C binding side?

Ptival committed:
first compiling draft of u-quark

Ptival committed:
forgot to update genPy

Ptival committed:
more principled way of generating handlers

now generates handlers for each component, for each message,
even in their absence.

fixes #12

Ptival committed:
generate unique names in ValidExchange

fixes Don's bug for micro-ssh

Ptival committed:
made u-quark closer to zach's proposal

Ptival committed:
merged master into u-quark

Ptival commented:
@ztatlock Just merged master into u-quark, and it solved my passfd problem, great!

@d1jang Commit e8e0bc3 fixes your issue, but commit ad28f83 is in a cleaner state. I suggest you rebase your u-ssh branch onto the latter (which, at the moment, is where the u-quark branch points to). If you feel incomfortable about whether and how to do this, we can go over it tomorrow.

It is now time to make u-quark work!

Ptival committed:
a minimal u-quark works!

Ptival committed:
merged master into u-quark, templates included

Ptival committed:
followed @ztatlock 's comment

Ptival committed:
merged master into u-quark

ztatlock commented on compiler/src/genCoq.ml:
```diff
@@ -17,16 +17,19 @@ let coq_of_typ = function
   | Num -> "num"
   | Str -> "str"
   | Fdesc -> "fdesc"
+  | Chan -> "tchan"
```
Ideally, adding chans could have been a separate PR, but no big deal.

ztatlock commented on compiler/src/genCoq.ml:
```diff
@@ -281,6 +284,29 @@ let coq_of_msg_pat m =
+    let compare = compare
+  end
+
+module ChanSet = Set.Make(OrderedChan);;
```
nice

ztatlock commented on compiler/src/genCoq.ml:
```diff
@@ -324,12 +358,24 @@ let coq_of_handler s xch_chan h =
+    let chans = chans_of_prog h.respond in
+    frames_of_vars s @ (
+      List.fold_left
+      (* folding function *)
```
Why not

``` ocaml
(fun acc elt -> mkstr "bound %s" elt :: acc)
```

(again, just a nitpick, not a big deal -- it makes it 1 line, then you prolly don't need comment? maybe...)

Ptival replied:
Yep right. At some point it was: 

``` OCaml
mkstr "bound %s" elt
:: mkstr "[In %s comps]" elt
:: acc

or something even longer maybe, so I made lines, but now that it's simple again, your point is valid.
```

ztatlock commented:
Looks pretty good to me.

Is the next step to merge master into this branch so that the merge can be accomplished automatically?

Ptival commented:
Ready to merge into master @ztatlock ?

ztatlock commented:
Merge at will @Ptival !
